src/lib/scan-imports.ts

```diff
@@ -0,0 +1,148 @@
+  const [_, webModule] = noExt.split(lastPart);
+  const moduleName = webModule || noExt; // if --dest wasn’t present, take the whole name
+
+  // return the module parts that should map to deps or devDeps
+  const [scope, pkg] = explode(moduleName);
+  const isScoped = scope.startsWith('@');
+  return isScoped ? `${scope}/${pkg}` : scope;
+}
+
+/**
+ * Return import name (or undefined)
+ */
+function importModuleName(node: NodePath<ImportDeclaration>): string | undefined {
+  if (isImportDeclaration(node)) {
+    return node.node.source.value;
+  }
+  return undefined;
+}
+
+/**
+ * Return dynamic import name (or undefined)
+ */
+function dynamicImportModuleName(node: NodePath<CallExpression>): string | undefined {
+  if (isCallExpression(node)) {
+    if (node.node.callee.type === 'Import' && node.node.arguments && node.node.arguments.length) {
+      return (node.node.arguments[0] as any).value;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Return array of dependencies
+ */
+function parseImports(file: string) {
+  const code = fs.readFileSync(file, 'utf8');
+
+  // skip file if empty, or if permissions error
+  if (!code) {
+    return [];
+  }
+  const deps = new Set<string>(); // micro-optimization: use Set() to dedupe
+
+  // try to parse; this will fail if not JS
+  try {
+    const ast = parse(code, {plugins: ['dynamicImport'], sourceType: 'module'});
+    traverse(ast, {
+      enter(node) {
+        const moduleName =
+          importModuleName(node as NodePath<ImportDeclaration>) ||
+          dynamicImportModuleName(node as NodePath<CallExpression>);
+
+        if (moduleName) {
+          deps.add(moduleName);
+        }
+      },
+    });
+    return [...deps]; // return Array for simplicity
+  } catch (e) {
+    return [];
+  }
+}
+
+interface ScanOptions {
+  dependencies: {[key: string]: string};
+  dest: string;
+}
+
+/**
+ * Scan files and their imports for dependencies
+ */
+export default function scanImports(entry: string, options: ScanOptions): string[] {
+  // return on missing entry
+  if (!entry.length) {
+    return [];
+  }
+
+  // if package.json has no dependencies, skip
+  if (!Object.keys(options.dependencies).length) {
+    return [];
+  }
+
+  // return & warn on no matching files
+  const files = glob.sync(entry, {nodir: true});
+  if (!files.length) {
+    console.warn(`No files found matching ${entry}`);
+    return [];
+  }
+
+  // start perf benchmark
+  const spinner = ora(`Scanning ${entry}`).start();
+  const timeStart = process.hrtime();
+
+  // get all dependencies within globs
+  const allDependencies = new Set<string>();
+  files.forEach(file => {
+    parseImports(file).forEach(dep => {
+      allDependencies.add(dep);
+    });
+  });
+
+  // filter out dependencies not in deps or devDeps
+  const npmDependencies = [...allDependencies]
+    .map(dep => getNpmName(dep, options.dest))
+    .filter(dep => !!options.dependencies[dep])
+    .sort((a, b) => a.localeCompare(b));
```

Comment:
So to sum up the other closed conversations, I just wanted to double-check on your thinking for this logic. This is the only thing outstanding, I think, and hopefully will answer your remaining Qs.

### Logic
1. Grabs all `import` and `import()`s in the glob files. Doesn’t care what they are.
2. If this is a web module installed by @pika/web, it attempts to figure out what its npm name should be (explained in comment above, but overall it’s pretty failsafe)
3. It then loops through all the module names, and if it’s in `dependencies` or `devDependencies`, it’ll add to the `arrayOfDeps()` array in `src/index`.

In my tests, this how it’s been resolving the following:

```
"/web_modules/vue/dist/vue.esm.browser.js"   -> "vue"
"/web_modules/d3.js"                         -> "d3"
"./pages/about.js"                           -> (ignored)
"./styles.css"                               -> (ignored)
"react"                                      -> "react"
```

Overall, though, I feel pretty confident about this logic. This shouldn’t ever fail to build; worst-case scenario if a user has some fancy rewrite rules; it just won’t automatically detect it and a user will have to add it manually to `webDependencies`.

Reply:
An alternative like we talked about: a `--prefix` flag (e.g.: `--prefix web_modules`) that will just remove that from the beginning of every module name in an attempt to resolve.

I’m not crazy about that just because:
- it’s more config
- unclear syntax (e.g. will `--prefix /web_modules/`, `--prefix /web_modules/`, and `--prefix web_modules` all work the same?)
- if a user changes `--dest`, and they’re also setting `--prefix`, they’d have to know to set that too (there wouldn’t be an error; they’d just have to notice modules were missing)
- `--prefix` probably won’t work if a user is using relative `../web_modules` and `../../web_modules` in their code

Adding a flag wouldn’t be logic-free; we’d need to add checks to handle these scenarios from all the different setups and user inputs we’d get. And I think it’d amount to more code than the little bit that’s there currently.

Anyway, I don’t mean to re-trace convos; I just keep coming back to what’s here and I feel like it’s a good solution.

Reply:
+1 to everything said above, i think the work you've done here strikes a good balance between smart and straightforward/simple. Agreed that we can hold off on a `--prefix` flag until we have a real example of someone hitting this issue.

Reply:
```
If this neither a web module nor a local file, assume it’s an npm package. Search for the module name as-is, but first check if it’s a submodule ( "/vue/dist/vue.esm.browser.js" -> "vue")
```

Can you say more about how this works? This is the only one I'm not sure that I understand. Does it just strip off the first `/` and then see if it exists in node_modules?

Reply:
Okay, I was able to run it locally and ran into this as an issue:

Certain packages need to be pointed at by the sub-path (where the package itself doesn't export a web-friendly file) but it appears that the current behavior is to strip sub-paths. For exampleIn my app I have `import algoliasearch from 'algoliasearch/dist/algoliasearchLite.js';`, and in the webDependencies I reference 'algoliasearch/dist/algoliasearchLite.js' directly, but this PR now tries to install 'algoliasearch' which is node.js specific and breaks the install.  Instead of stripping the subpath, we should keep it in since that's defined behavior in the README.

Reply:
Hm. So the issue I ran into is that if we don’t do some figuring to see if it’s an npm package or not, then we get issues when we pass local modules like `./about.js` to the `install()` function.

So would you propose in this instance, still validating that `algoliasearch` is an npm package, but pass `algoliasearch/dist/algoliasearchLite.js` to the `install()` function?

Reply:
Ohhhhkay I found exactly what you were talking about. Yup. That was an issue. I’m going to resolve this convo for cleanliness and comment on the line changes. Good catch!